File: crates/fusion-py/python/fusion_framework/cache_monitor.py
"""Built-in process-wide cache monitor panel (HTML + JSON).

Enabled only when ``cache.monitor.enabled`` is true in ``fusion.<env>.json``.
When disabled, routes are not registered (security: no endpoints, not just UI).
"""

from __future__ import annotations

from typing import Any

from fusion_framework.template import FusionBaseTemplate


class CacheMonitorPanel(FusionBaseTemplate):
    """Default cache monitoring page using Fusion Tera components."""

    template = "fusion/cache_monitor.html"

    def context(self) -> dict[str, Any]:
        """Live entries + recent set/delete/clear events for the panel."""
        from fusion_framework import cache

        return dict(cache.panel_context())


def _truthy_enabled(value: Any, default: bool = False) -> bool:
    """Interpret settings booleans (same rules as swagger.enabled)."""
    if value is None:
        return default
    if value in (False, 0, "false", "False", "0", "off", "no"):
        return False
    if value in (True, 1, "true", "True", "1", "on", "yes"):
        return True
    return bool(value)


def _normalize_monitor_path(raw: Any) -> str:
    """Normalize monitor URL path (leading slash, no trailing slash)."""
    path = str(raw or "/__fusion/cache").strip() or "/__fusion/cache"
    if not path.startswith("/"):
        path = f"/{path}"
    return path.rstrip("/") or "/__fusion/cache"


def mount_cache_monitor(engine, settings) -> bool:
    """Register HTML + ``/json`` routes when ``cache.monitor.enabled`` is true.

    Returns whether routes were mounted.
    """
    from fusion_framework import cache

    if not _truthy_enabled(settings.get("cache.monitor.enabled", default=False), default=False):
        return False

    # Apply settings so monitor_path / max_events match the env file.
    cache.configure(settings)
    path = _normalize_monitor_path(
        settings.get("cache.monitor.path", default="/__fusion/cache")
    )

    def html_handler(req: dict) -> Any:
        """Serve the monitor HTML (or JSON when Accept prefers JSON)."""
        return CacheMonitorPanel(req).get()

    def json_handler(_req: dict) -> Any:
        """Serve the raw monitor snapshot JSON."""
        return cache.snapshot()

    engine.route("GET", path, html_handler)
    if path != "/":
        engine.route("GET", f"{path}/", html_handler)
    engine.route("GET", f"{path}/json", json_handler)
    return True


__all__ = ["CacheMonitorPanel", "mount_cache_monitor"]

File: bindings/csharp/FusionFramework/CacheMonitor.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace FusionFramework;

/// <summary>
/// Built-in process-wide cache monitor (HTML + JSON).
/// Mounted only when <c>cache.monitor.enabled</c> is true — disabled means no endpoints.
/// </summary>
public static class CacheMonitor
{
    /// <summary>Register HTML and <c>/json</c> routes when enabled in settings.</summary>
    public static bool Mount(FusionApp app, FusionSettings settings)
    {
        if (!Truthy(settings.Get("cache.monitor.enabled", false), false))
            return false;

        Cache.Configure(settings);
        var path = NormalizePath(AsString(settings.Get("cache.monitor.path", "/__fusion/cache")));

        app.AddRawRoute("GET", path, () => new CacheMonitorPanel().Get());
        if (path != "/")
            app.AddRawRoute("GET", $"{path}/", () => new CacheMonitorPanel().Get());
        app.AddRawRoute("GET", $"{path}/json", () => Cache.Snapshot());
        return true;
    }

    static string NormalizePath(string? raw)
    {
        var path = string.IsNullOrWhiteSpace(raw) ? "/__fusion/cache" : raw.Trim();
        if (!path.StartsWith('/')) path = "/" + path;
        path = path.TrimEnd('/');
        return string.IsNullOrEmpty(path) ? "/__fusion/cache" : path;
    }

    static bool Truthy(JsonNode? value, bool defaultValue)
    {
        if (value is null || value.GetValueKind() == JsonValueKind.Null)
            return defaultValue;
        if (value.GetValueKind() == JsonValueKind.False) return false;
        if (value.GetValueKind() == JsonValueKind.True) return true;
        if (value is JsonValue v)
        {
            if (v.TryGetValue<bool>(out var b)) return b;
            if (v.TryGetValue<int>(out var n)) return n != 0;
            if (v.TryGetValue<string>(out var s))
            {
                return s.ToLowerInvariant() switch
                {
                    "false" or "0" or "off" or "no" => false,
                    "true" or "1" or "on" or "yes" => true,
                    _ => !string.IsNullOrWhiteSpace(s),
                };
            }
        }
        return true;
    }

    static string? AsString(object? value)
    {
        if (value is null) return null;
        if (value is string s) return s;
        if (value is JsonValue jv && jv.TryGetValue<string>(out var str)) return str;
        if (value is JsonNode node)
            return node.GetValueKind() == JsonValueKind.String ? node.GetValue<string>() : node.ToJsonString();
        return value.ToString();
    }
}

/// <summary>Default cache monitor page (FusionBaseTemplate + built-in components).</summary>
public sealed class CacheMonitorPanel : FusionBaseTemplate
{
    /// <inheritdoc />
    public override string TemplateName() => "fusion/cache_monitor.html";

    /// <inheritdoc />
    public override Dictionary<string, JsonNode?> Context()
    {
        var node = Cache.PanelContext();
        var dict = new Dictionary<string, JsonNode?>(StringComparer.Ordinal);
        if (node is JsonObject obj)
        {
            foreach (var kv in obj)
                dict[kv.Key] = kv.Value?.DeepClone();
        }
        return dict;
    }
}
